### PR TITLE
Add SMCFilter class; Add simple model example that uses SMC Filter

### PR DESCRIPTION
**Pull request by chanjed**
Addresses #1304 
pair coded with @fritzo and @karalets 

This PR:
1. Implements SMCFilter class
2. Simple State Space model example to run SMCFiltering on.

## Tasks
- [x] add examples/smc.py
- [x] add a file tutorial/source/smc.rst following the pattern of tutorial/source/hmm.py, and register tutorial/source/smc.rst in tutorial/source/index.rst
- [x] add a docstring to `SMCFilter` as in other inference algorithms
- [x] add an `SMCFilter` section to docs/source/inference_algos.rst
- [x] add smoke tests with different numbers of sample sites and different `event_shape` and `batch_shape`. Tests should live in tests/infer/test_smcfilter.py
- [x] add a simple statistical test that evaluates correctness

## Traiged
- Implement Conjugate SMC Class (this should be in a separate PR)

### CODE REVIEW

File: examples/smc.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+from __future__ import absolute_import, division, print_function
+
+import argparse
+
+import torch
+
+import pyro
+import pyro.distributions as dist
+import pyro.infer.smcfilter as smcfilter
```
Comment (fritzo) [inline examples/smc.py]: nit: can you `from pyro.infer import SMCFilter` and use `SMCFilter` below?

File: examples/smc.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+from __future__ import absolute_import, division, print_function
```
Comment (fritzo) [inline examples/smc.py]: Can you add a little docstring at the top of this file, similar to those in other examples. Something like
```py
"""
This file demonstrates how to use the SMCFilter algorithm with a simple model of a noisy harmonic oscillator.
"""
```

**Comment by tuananhle7** [comment]
Hi @chanjed, @fritzo and @karalets,

Thank you for implementing this feature.

I was looking at [this example](https://github.com/pyro-ppl/pyro/blob/dev/examples/smcfilter.py) and the [source code](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/smcfilter.py) in order to understand the implementation.

I don't understand how resampling works:

Assume that at time `t - 1`, we have particles `{z_(t - 1)k}` and weights `{w_(t - 1)k}` (`k = 1, ..., K`) which I think (?) are stored in [`_values`](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/smcfilter.py#L42) and [`_log_weights`](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/smcfilter.py#L43).

The [resampling step](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/smcfilter.py#L73) produces ancestral indices `a_(t - 1)k` (`k = 1, ..., K`) which are marginally distributed according to a Categorical distribution parameterized by the normalized weights ([here](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/smcfilter.py#L123) I think). Then `z_(t - 1)k` which is stored in `_values` is updated to be `z_(t - 1)(a_(t - 1)k)` [here](https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/smcfilter.py#L124).

At time `t`, we should propose `z_tk` from `guide(z_t(k - 1))` where `z_t(k - 1)` is already resampled. However, [here](https://github.com/pyro-ppl/pyro/blob/dev/examples/smcfilter.py#L60-L61), `z_tk` is proposed from a guide which is conditioned on `self.z` (attribute of guide) which doesn't seem to interact with `_values` which stores the resampled values.

Thank you,
Tuan Anh
